[router][controller][server] Cleanups to partition/replica state

FelixGV opened the pull request:
Various parts of the code dealt with status in the form of strings, and those strings corresponded to values of the ExecutionStatus or HelixState enums. Worse, in some cases, a string could represent either of these enums. This commit fixes this by using strong types instead.

- Partition is the most significantly refactored class, leading to the elimination of the string-keyed map of states to host list. The old constructor that takes in a string-keyed map is still there but only used by tests and marked deprecated. All main code paths now construct instances via a new more efficient constructor that takes in EnumMaps directly.

- ReplicaState now carries just an ExecutionStatus rather than both states, each in a separate string. Also removed the isReadyToServe field, since it could be inferred from the venicePushStatus instead.

- PartitionStatus removes some redundant map lookups.

- ReplicaStatus now avoids processing statusHistory in code paths that don't need it.

- HelixExternalViewRepository::getReplicaStates now throws an exception. It was never used in the previous code, and carried lingering uses of code paths that are now cleaned up.

- HelixCustomizedViewOfflinePushRepository::onCustomizedViewDataChange has a few efficiency improvements related to Partition states and also reduces logging volume in cases instances were found in the CV, but missing from Live Instances. This is now going to be printed once per ZK watch invocation, with all relevant hostnames included, rather than once per missing replica.

- Deleted RoutingDataRepository::refreshRoutingDataForResource and all implementations.


## How was this PR tested?
GHCI

## Does this PR introduce any user-facing changes?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, choose 'No'.
-->
- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

sixpluszero commented on internal/venice-common/src/test/java/com/linkedin/venice/pushmonitor/PartitionStatusTest.java:
```diff
@@ -1,5 +1,7 @@
 package com.linkedin.venice.pushmonitor;
 
+import static org.testng.Assert.*;
```
nit: Let's not use * import here.

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -26,19 +33,33 @@
   private static final Logger LOGGER = LogManager.getLogger(Partition.class);
   /**
-   * ID of partition. One of the number between [0 ~ total number of partition)
+   * ID of partition. One of the number between [0, total number of partition[
```
maybe ] instead of [

sixpluszero replied:
Hi Felix, closing this issue as Sushant also posted the same one. I might agree what he said - original syntax might be correct?

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -102,13 +128,23 @@ public List<Instance> getInstancesInState(String state) {
   public List<Instance> getReadyToServeInstances() {
```
Unrelated to your change but ... the JavaDoc says remove this API once we are fully on L/F.. do you think we could remove it or it is the doc has problem?

FelixGV replied:
I cleaned this up.

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -102,13 +128,23 @@ public List<Instance> getInstancesInState(String state) {
+    return completed.size() > online.size() ? completed : getReadyInstances();
   }
 
   public List<Instance> getWorkingInstances() {
```
Also, just to be cautious as I am confused by the existing variable/API naming here. Is working instance containing the ready-to-serve and bootstrapping ones?

FelixGV replied:
Yeah... although I think that's a hangover from the O/B/O days, since we don't even have `BOOTSTRAP` anymore... so most likely, `getWorkingInstances()` could be substituted for `getReadyToServeInstances()` and it would work just as well...

FelixGV replied:
I ended up cleaning this up. Deeeeeep rabbit hole. But I think with this we'll be completely rid of O/B/O. I even deleted `ONLINE` and `BOOTSTRAP` from the `HelixState` enum!

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -131,52 +167,96 @@ public Instance getLeaderInstance() {
+        }
+      }
+    }
+    return null;
```
Should we return HelixState.UNKNOWN instead?

sixpluszero replied:
Or here Unknown has different meaning than not found instance id...

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/pushmonitor/PushStatusDecider.java:
```diff
@@ -251,42 +252,29 @@ private void logConditionally(String newReason, Optional<String> previousReason,
       ExecutionStatus currentStatus =
           getReplicaCurrentStatus(partitionStatus.getReplicaHistoricStatusList(entry.getKey().getNodeId()));
-      if (entry.getValue().equals(HelixState.LEADER_STATE)) {
+      if (entry.getValue() == HelixState.LEADER) {
```
Hmm, I thought .equals() should be the better way?

majisourav99 replied:
enum `==` projects against NPE.

FelixGV replied:
Enum values are singletons, so we can do `==` on them, just like with primitive.

I prefer doing `==` whenever possible since it is more concise (and for me, easier to read), it gets compile-time type safety (whereas you could call `Enum1.BLAH.equals(Enum2.BLEH)` and technically it compiles), and you can't get NPEs on `==`.

sixpluszero commented on internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:
```diff
@@ -47,42 +49,7 @@ public HelixExternalViewRepository(SafeHelixManager manager) {
   }
 
   public List<ReplicaState> getReplicaStates(String kafkaTopic, int partitionId) {
-    Partition partition = resourceAssignment.getPartition(kafkaTopic, partitionId);
```
not in the scope of this PR: This change looks good for this cleanup PR but I'd suggest we move the API to CV based repo only and force MetaDataHandler to instantiate HelixCVRepo instead general repo.

FelixGV replied:
SG, done.

sixpluszero reviewed (commented):
Thanks for the cleanup. I took a pass and left some comments. Also, it looks like HelixState.java has some xxxx_STATE which can all be replaced by enum.name()?

sushantmane commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -26,19 +33,33 @@
   private static final Logger LOGGER = LogManager.getLogger(Partition.class);
   /**
-   * ID of partition. One of the number between [0 ~ total number of partition)
+   * ID of partition. One of the number between [0, total number of partition[
```
IMO original syntax was correct. It is half-closed interval.

`[0, total number of partition) --> [a, b) ==> {x belons to N | a <= x && x < b}`

FelixGV replied:
Seems like both syntaxes are correct... I didn't know about the parenthesis syntax for exclusion!

https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints

sushantmane commented on internal/venice-common/src/main/java/com/linkedin/venice/pushmonitor/PushStatusDecider.java:
```diff
@@ -63,14 +64,16 @@ public Pair<ExecutionStatus, Optional<String>> checkPushStatusAndDetails(
-          assignedReplicasCount,
-          errorReplicasCount,
-          completedReplicasCount);
+      if (logger.isDebugEnabled()) {
```
Is this check necessary if we use lambda for ` partition.getNumOfTotalInstances()`?

FelixGV replied:
Lambda could be an alternative way to do it, but it would instantiate a lambda object, which isn't necessary IMHO...

sushantmane reviewed (commented):
Thanks, @FelixGV! I left a few comments, please take a look.

FelixGV commented on internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixExternalViewRepository.java:
```diff
@@ -47,42 +49,7 @@ public HelixExternalViewRepository(SafeHelixManager manager) {
   }
 
   public List<ReplicaState> getReplicaStates(String kafkaTopic, int partitionId) {
```
Instead I followed Jialin's advice to remove this from the interface and leave it only inside `HelixCustomizedViewOfflinePushRepository`.

FelixGV commented on internal/venice-common/src/main/java/com/linkedin/venice/meta/Partition.java:
```diff
@@ -26,19 +33,33 @@
   public Partition(int id, Map<String, List<Instance>> stateToInstancesMap) {
     this.id = id;
-    this.stateToInstancesMap = stateToInstancesMap;
 
```
I completely eliminated the deprecated constructor in my latest diff.

FelixGV commented:
> Thanks for the cleanup. I took a pass and left some comments. Also, it looks like HelixState.java has some xxxx_STATE which can all be replaced by enum.name()?

This is not possible... the annotations require a fully static constant, which apparently does not allow calling this function...

sixpluszero commented on internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controller/TestHAASController.java:
```diff
@@ -87,7 +87,7 @@ public void testStartHAASHelixControllerAsControllerClusterLeader() {
   }
 
-  @Test(timeOut = 60 * Time.MS_PER_SECOND)
+  @Test(timeOut = 120 * Time.MS_PER_SECOND)
```
Is this for mitigating flaky test?

FelixGV replied:
Yeah, I was getting flakiness on these two tests on my local, and increasing the timeout seemed to help.

sixpluszero commented on internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controller/TestHAASController.java:
```diff
@@ -125,6 +125,7 @@ public void testTransitionToHAASControllerAsControllerClusterLeader() {
               () -> assertEquals(
                   assertCommand(controllerClient.queryJobStatus(topicName)).getStatus(),
                   ExecutionStatus.COMPLETED.toString())));
+      System.out.println("End of the test");
```
Same here, but since this it test, probably fine to keep it (probably for the sake of debugging flakiness..)

FelixGV replied:
Thanks for catching it... I thought I had removed it, but forgot. Will clear it in the next commit.

sixpluszero reviewed (approved):
LGTM from my side! Thanks for cleaning up legacy implementations.

FelixGV commented:
I have rebased and pushed one more commit, which removes the unnecessary log @sixpluszero pointed out, as well as some dead code I found. I also rewrote the commit message to include all details from the various iterations. Thanks for the reviews!

sixpluszero reviewed (approved):
LGTM thanks!

FelixGV commented:
> LGTM thanks!

Thanks for the review!
